designs/2019-move-to-async-api/README.md

```diff
@@ -0,0 +1,531 @@
+
+```js
+const { ESLint } = require("eslint")
+const eslint = new ESLint()
+const formatter = eslint.getFormatter("stylish")
+
+// Verify files
+let results = eslint.executeOnFiles(patterns)
+// Update the files of the results if needed
+if (process.argv.includes("--fix")) {
+    results = ESLint.outputFixesInIteration(results)
+}
+// Format and write the results
+for await (const textPiece of formatter(results)) {
+    process.stdout.write(textPiece)
+}
+```
+
+#### ● The `extractErrorResults()` method
+
+The original `CLIEngine.getErrorResults()` static method receives an array of lint results, then extracts only the messages, which are error severity, then returns the results that contain only those.
+
+This method just changed the arrays to async iterables because this method is sandwiched between `executeOnFiles()` and formatters.
+
+```js
+const { ESLint } = require("eslint")
+const eslint = new ESLint()
+const formatter = eslint.getFormatter("stylish")
+
+// Verify files
+let results = eslint.executeOnFiles(patterns)
+// Extract only error results
+if (process.argv.includes("--quiet")) {
+    results = ESLint.extractErrorResults(results)
+}
+// Format and write the results
+for await (const textPiece of formatter(results)) {
+    process.stdout.write(textPiece)
+}
+```
+
+#### ● The other methods
+
+The following methods return `Promise` which gets fulfilled with each result. Once we got this change, we can support ES modules for shareable configs, plugins, and custom parsers without more breaking changes.
+
+- `getConfigForFile()`
+- `getRules()`
+- `isPathIgnored()`
+
+The following methods are removed because those don't fit the new API.
+
+- `addPlugin()` ... This method has caused to confuse people. We have introduced this method to add plugin implementations and expected people to use this method to test plugins. But people have often thought that this method loads a new plugin for the following linting so they can use plugins rules without `plugins` setting. And this method is only one that mutates the state of `CLIEngine` objects and messes all caches. Therefore, this RFC moves this functionality to a constructor option. See also "[Constructor](#-constructor)" section.
+- `resolveFileGlobPatterns()` ... ESLint doesn't use this logic since `v6.0.0`, but it has stayed there for backward compatibility. Once [RFC20] is implemented, what ESLint iterates and what the glob of this method iterates will be different, then it will confuse users. This is good timing to remove the legacy.
+
+#### ● New methods
+
+- `collectResults()` ... This method receives an async iterable object then returns an array that contains the iterated values.
```

Comment:
This doesn't seem ESLint specific?

Reply:
Right, this is not ESLint specific.

Reply:
In that case, I'm not sure it's worth being part of the API.